frankDoc/src/main/java/nl/nn/adapterframework/frankdoc/doclet/FrankMethod.java

```diff
@@ -28,4 +30,12 @@
 	FrankType[] getParameterTypes();
 	boolean isVarargs();
 	FrankAnnotation getAnnotationInludingInherited(String name) throws FrankDocException;
+	String getJavaDoc();
+	String getJavaDocIncludingInherited() throws FrankDocException;
+	String getDefaultValueFromJavadoc();
+	String getDefaultValueFromJavadocIncludingInherited() throws FrankDocException;
```

Comment:
Wat doen de `IncludingInherited` methodes precies? Zijn die nodig in dit interface? Moet hij niet altijd op zoek naar de inherited value (en deze overschijven wanneer er een lokale value is)?

Reply:
We voegen attributes en config children alleen toe aan een FrankElement als ze in de betreffende Java class gedefiniëerd zijn. Daarom is een method die alleen de declared Javadoc teruggeeft relevant. Maar als een attribute of config child gedefiniëerd is zonder Javadoc, dan willen we de inherited Javadoc gebruiken. Dit systeem hadden we al voor IbisDoc annotaties. Ik heb het nagedaan voor JavaDocs.

Reply:
Niels, je vroeg me door de telefoon waarom FrankMethod niet gewoon de default waarde bepaalt en dus al kiest of de default uit de @IbisDoc annotatie komt of uit de JavaDoc tags. Die keuze wordt nu pas gemaakt in het model. Deze class hoort niet tot het model, maar tot de wrapper API.

Het antwoord op deze vraag volgt uit de historie. Het Frank!Doc project begon ermee dat we een model maakten dat gebruikt werd door de classes die de XSDs en de JSON schreven. Het model verzamelde informatie als: welke elementen zijn er? wat zijn hun attributes? wat zijn de default values van de attributes?, enz. Die informatie werd uitsluitend bepaald door Java reflectie omdat we nog geen doclets gebruikten. Het was toen logisch dat het model de default waardes van attributes uitrekende. Het model gebruikte Java reflectie om een IbisDoc annotatie op te halen en parste de waardes die erin zaten.

Daarna gingen we doclets gebruiken. Toen pas hebben we de wrapper API gemaakt met interfaces als FrankClass en FrankMethod. We wilden geen dubbel werk doen in de wrapper API en in het model. Daarom moest de wrapper API dezelfde soort functionaliteit bieden als Java reflectie of de doclet API. Dus gewoon ophalen welke classes, methods en annotaties er zijn, en de interpretatie van die informatie overlaten aan het model. De informatie: "wat is de default waarde van een attribute?" wordt verkregen door annotaties (nu zowel Java 5 annotaties als JavaDoc annotaties) te interpreteren.

Ik stel voor dat we de methods getDefaultValueFromJavadoc() en getDefaultValueFromJavadocIncludingInherited() vervangen door methods getJavadocTag(String tagName) en getJavadocTagIncludingInherited(String tagName). We zouden nog de bestaande methods getAnnotations(), getAnnotation(String name) en getAnnotationIncludingInherited(String name) kunnen hernoemen naar getJava5Annotations(), getJava5Annotation(String name) en getJava5AnnotationIncludingInherited(String name). Deze updates maken weer duidelijk hoe de verantwoordelijkheden tussen de wrapper API en het model verdeeld zijn.

Het invoeren van getJava5xxx namen kunnen we denk ik het beste in een aparte pull request doen. Dit is gewoon Eclipse replace. Als dit in een aparte pull request gebeurt, dan hoeft die pull request nauwelijks gereviewed te worden.